src/vm/functions/conversions.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use vm::costs::cost_functions::ClarityCostFunction;
+use vm::costs::runtime_cost;
+use vm::errors::{check_argument_count, CheckErrors, InterpreterResult as Result};
+use vm::representations::SymbolicExpression;
+use vm::types::BufferLength;
+use vm::types::SequenceSubtype::{BufferType, StringType};
+use vm::types::StringSubtype::ASCII;
+use vm::types::TypeSignature::SequenceType;
+use vm::types::{SequenceData, TypeSignature, Value};
+use vm::{apply, eval, lookup_function, Environment, LocalContext};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum EndianDirection {
+    LittleEndian,
+    BigEndian,
+}
+
+// The functions in this file support conversion from (buff 16) to either 1) int or 2) uint,
+// from formats 1) big-endian and 2) little-endian.
+//
+// The function 'buff_to_int_generic' describes the logic common to these four functions.
+// This is a generic function for conversion from a buffer to an int or uint. The four
+// versions of Clarity function each call this, with different values for 'conversion_fn'.
+//
+// This function checks and parses the arguments, and calls 'conversion_fn' to do
+// the specific form of conversion required.
+pub fn buff_to_int_generic(
+    value: Value,
+    direction: EndianDirection,
+    conversion_fn: fn([u8; 16]) -> Value,
+) -> Result<Value> {
+    match value {
+        Value::Sequence(SequenceData::Buffer(ref sequence_data)) => {
+            if sequence_data.len() > BufferLength(16) {
+                return Err(CheckErrors::TypeValueError(
+                    SequenceType(BufferType(BufferLength(16))),
+                    value,
+                )
+                .into());
+            } else {
+                let mut buf = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+                let mut original_slice = sequence_data.as_slice().to_vec();
```

Comment:
You might be able to get away with doing an internal `malloc()` here if you instead copied the original slice directly into `buf` (no need to reverse it), and passed _that_ into the relevant `::from_be_bytes()` or `::to_be_bytes()` methods.

Reply:
I made some changes to maybe make it easier to understand. And, we avoid the 'reverse'.